### PR TITLE
tuple type introspection via indexing/iteration

### PR DESCRIPTION
**Pull request by stevengj**
This PR simplifies introspection of tuple types by allowing you to iterate/index tuple types almost as if they were simply a tuple of types.  For example, it allows you to do:

```jl
julia> T = Tuple{Int8,String,Float32}
Tuple{Int8,String,Float32}

julia> length(T)
3

julia> T[2]
String

julia> collect(Type, T)
3-element Array{Type,1}:
 Int8
 String
 Float32
```

Without this PR, you have to do a lot of digging through Julia internals to do introspection of tuple types (via the undocumented `T.types` field), and it is especially difficult to introspect vararg tuple types (via undocumented functions like `Base.unwrapva`).

I encountered this when generalizing tuple conversion for PyCall (JuliaPy/PyCall.jl#404), and the difficulty of tuple introspection has also caused some fragility in JLD (#10999).

### CODE REVIEW

**Comment by quinnj** [comment]
+1, I've definitely wanted this lately.

**Comment by stevengj** [comment]
See also #22350.

File: base/tuple.jl [hunk 1]
```diff
@@ -310,3 +310,23 @@ any(x::Tuple{}) = false
 any(x::Tuple{Bool}) = x[1]
 any(x::Tuple{Bool, Bool}) = x[1]|x[2]
 any(x::Tuple{Bool, Bool, Bool}) = x[1]|x[2]|x[3]
+
+## tuple-type introspection as array/iterator ##
+# ... we can't define iteratorsize/iteratoreltype, however, because
+#     typeof(Tuple) is simply DataType
+_iteratorsize(::Type{T}) where {T<:Tuple} = (@_pure_meta; isvatuple(T) ? IsInfinite() : HasLength())
+_length(::HasLength, T::Type{<:Tuple}) = length(T.types)
+_length(::IsInfinite, T::Type{<:Tuple}) = typemax(Int)
```
Comment (tkelman) [inline base/tuple.jl]: little odd for something that's supposed to be infinite to return a finite length?

**Comment by stevengj** [comment]
It also might be useful to export a new function like `isvararg(T::Tuple)` that returns `true` for `Vararg` tuple types.

File: base/tuple.jl (hunk 1 above)
Comment (stevengj) [inline base/tuple.jl]: Should it throw an error instead?   (I could just delete this line and it would throw a `MethodError` in that case.)

Comment (JeffBezanson) [inline base/tuple.jl]: Yes, `length` usually gives a MethodError on infinite iterators.

**Comment by stevengj** [comment]
One problem here is that `Union{Tuple{Any,Int},Tuple{Any}} <: Tuple` (see also tonyhffong/Lint.jl#215), so the current version of this PR will do the wrong thing on unions of tuples.  

Ideally, I'd like to be able to introspect the components of unions in the same way, so adding some additional methods to this PR would do the trick.  However, I'm confused by the behavior of dispatch here:
```jl
julia> foo(::Type{<:Union}) = 1
       foo(::Type{<:Tuple}) = 2
       foo(::Type{Union{<:Tuple}}) = 3

julia> foo(Union{Tuple{Int},Tuple{Any}})
2
```
I would have expected `1` or `3`, not `2` ... how can you dispatch on a union of tuples?

**Comment by mbauman** [comment]
Ref https://github.com/JuliaLang/julia/pull/11547.

**Comment by vtjnash** [comment]
In https://github.com/JuliaLang/julia/issues/22051, I observed that while accessing the fields of the tuple datatype is generally not valid (because of the possibility of Union), and thus also not possible to infer, the `fieldtype` function may be a solution. It should be dealing with those other possibilities correctly anyways, so it already largely must have the right semantics.

**Comment by vtjnash** [comment]
> foo(::Type{Union{<:Tuple}})

This is just `Type{Tuple}` and matches precisely `Tuple`

> foo(::Type{<:Union}) = 1

This is just `Type{Union}` and matches precisely `Union`

`foo(::Union) = 1` might work, but dispatch to it may not be reliable (I can't find the right issue number just now), and I'm surprised at this sort order, since kinds usually are expected to be sorted after Types:

```
julia> methods(foo)
# 6 methods for generic function "foo":
[1] foo(::Core.TypeofBottom) in Main at REPL[9]:1
[2] foo(::Union) in Main at REPL[2]:1
[3] foo(::Type{Union{}}) in Main at REPL[8]:1
[4] foo(::Type{#s1} where #s1<:Tuple) in Main at REPL[6]:1
[5] foo(::DataType) in Main at REPL[1]:1
[6] foo(::UnionAll) in Main at REPL[4]:1
```

**Comment by stevengj** [comment]
@vtjnash, thanks for the tip about `fieldtype`.   It seems to work, including the vararg case. 
 e.g. `fieldtype(Tuple{Int, Vararg{String}}, i)` returns `Int` for `i==1` and `String` for `i≥1`.

In order to use this effectively, however, it would be nice to export and document some version of `Base.isvatuple`, however, and have a documented way to get the number of fields other than `length(fieldnames(T))`.

**Comment by JeffBezanson** [comment]
Yes, in this case it's important to distinguish looking at the structure of a type (which is generally painful and difficult to get right) from "type-domain indexing", which is what `getfield_tfunc` computes:

```
julia> Core.Inference.getfield_tfunc(Union{Tuple{Int},Tuple{String}}, Int)
Union{Int64, String}
```

That's telling you what you'd get from accessing an instance of the given type.

**Comment by martinholters** [comment]
While a better introspection API sounds useful, I'm not sure I like the `getindex` approach here:
```julia
julia> Int[1]
1-element Array{Int64,1}:
 1

julia> Tuple{Int}[(1,)]
1-element Array{Tuple{Int64},1}:
 (1,)

julia> Tuple{Int}[1] # surprise!
Int64
```
Putting this behind `fieldtype` has less potential for confusion, I'd say. If iteration is needed, maybe a new function`fieldtypes` could yield an iterable (and indexable) wrapper?

**Comment by andyferris** [comment]
Anything that makes working with these easier is very welcome to me. :)

Would a `tfunc` be necessary to make `getindex` return types inferred, or does this work already?

**Comment by stevengj** [comment]
I'm not so concerned about iteration (I can always loop over indices), but I am concerned with knowing the length of the tuple, whether it is a vararg tuple, and documenting this introspection technique; having multiple packages looking at undocumented internals is broken.

It would also be nice to make sure these things were known at compile time, but this is an optimization that could be added later if needed.

I will close this PR since the consensus seems to be against indexing (and I wasn't even aware of #11547).  Will try again with another PR based on `fieldtype`.

**Comment by bramtayl** [comment]
This works nicely:

```
@generated function tuple_types(tuple_type)
    Expr(:tuple, map(1:nfields(test)) do i
        :(fieldtype($test, $i))
    end...)
end
```

**Comment by vtjnash** [comment]
Or more succinctly, `inner_types(atype::Type) = ntuple(i -> fieldtype(atype, i), Val(nfields(atype)))` (provided, of course, that you are only calling this on a leaftype, and not something else like a Vararg Tuple)

**Comment by bramtayl** [comment]
Cool. Is there a way to get fieldtype to inline?

**Comment by bramtayl** [comment]
Boiled down the type instability to this:

```
atype = Tuple{Int64, String}
    
test(f, a) = f(a, 1), f(a, 2)
test2(f, a) = begin
    f2(i) = f(a, i)
    f2(1), f2(2)
end

Test.@inferred test(fieldtype, atype)
Test.@inferred test2(fieldtype, atype) # fails
```

**Comment by TotalVerb** [comment]
@martinholters I'd argue that it's really the `T[x, y]` syntax that's problematic there (as it's a pun for `getindex`), but I guess that's too late to change...

**Comment by bramtayl** [comment]
This should be fully inferable:

```
inner_value(::Val{T}) where T = T
function inner_types(atype)
    const_type = Val{atype}()
    Base.@pure test(i) = fieldtype(inner_value(const_type), i)
    ntuple(test, Val{nfields(atype)}())
end
```

**Comment by bramtayl** [comment]
Safer:

```
function inner_types(atype)
    const_type = Val{atype}()
    Base.@pure inner_function(i) = Val{fieldtype(inner_value(const_type), i)}()
    ntuple(inner_function, Val{nfields(atype)}())
end
```

I'm gathering that if you are going to use types in inference, they should be passed around wrapped in Vals, and unpacked at point of use with `inner_value`.

**Pull request closed without merging**